pageserver/src/tenant/timeline.rs

```diff
@@ -3505,12 +3506,27 @@ impl Timeline {
         // min-heap (reserve space for one more element added before eviction)
         let mut heap: BinaryHeap<Hole> = BinaryHeap::with_capacity(max_holes + 1);
         let mut prev: Option<Key> = None;
-        for (next_key, _next_lsn, _size) in itertools::process_results(
-            deltas_to_compact
-                .iter()
-                .map(|l| -> Result<_> { Ok(l.load_keys(ctx)?.into_iter()) }),
-            |iter_iter| iter_iter.kmerge_by(|a, b| a.0 < b.0),
-        )? {
+
+        let mut all_value_refs = Vec::new();
+        for l in deltas_to_compact.iter() {
+            // TODO: replace this with an await once we fully go async
+            all_value_refs.extend(Handle::current().block_on(l.load_val_refs(ctx))?);
+        }
+        // The current stdlib sorting implementation is designed in a way where it is
+        // particularly fast where the slice is made up of sorted sub-ranges.
+        all_value_refs.sort_by_key(|(key, _lsn, _value_ref)| *key);
```

Comment:
The kmerge documentation of itertools didn't write anything about it, but I think typically heap implementations are not stable, that's also why heapsort is not stable (typically). See also [this thread](https://cstheory.stackexchange.com/q/593). So yeah we could use `sort_unstable_by_key` but the question is if it's faster. The `sort_by_key` docs:

> The current algorithm is an adaptive, iterative merge sort inspired by [timsort](https://en.wikipedia.org/wiki/Timsort). It is designed to be very fast in cases where the slice is nearly sorted, or consists of two or more sorted sequences concatenated one after another.

While the `sort_unstable_by_key` function uses pdqsort, where [the docs say](https://github.com/orlp/pdqsort#the-worst-case):

> Quicksort naturally performs bad on inputs that form patterns, due to it being a partition-based sort. Choosing a bad pivot will result in many comparisons that give little to no progress in the sorting process. If the pattern does not get broken up, this can happen many times in a row.

> A bad partition occurs when the position of the pivot after partitioning is under 12.5% (1/8th) percentile or over 87,5% percentile - the partition is highly unbalanced. 

Our slice is lots of sorted sub-sets concatenated to each other, so this might be a pattern that the docs speak about. The sorting algorithm switches to heapsort if too many bad partitions are encountered.

Of course, it might be best to benchmark on real-world data to compare the two algorithms.

Reply:
> best to benchmark

Sadly I don't think we have a benchmark on compaction... We most certainly could have, at least an ugly kind which would require a lot of work to reset.

Perhaps this will be insignificant compared to the wins of sorting and doing it only once.